### PR TITLE
[ROCm] Fix warpMergeSortTopK padding sentinel for integer dtypes

### PR DESCRIPTION
**Pull request by lfmeadow**
## Summary

`warpMergeSortTopK` (introduced by #170029) uses `std::numeric_limits<scalar_t>::infinity()` (negated for descending) as the out-of-band sentinel for the slots that pad a slice out to the warp's `items_per_thread * WARP_SIZE` sort width. For integer `scalar_t`, `numeric_limits<T>::infinity()` returns `T()` i.e. `0`, which is a legal input value, not an out-of-band sentinel. The padding slots therefore sort among real input values and their placeholder indices (in `[inputSliceSize, sort_size)`) leak into the top-k output, producing out-of-bounds indices and mismatched values.

This dispatch path is currently only enabled on ROCm ≥ 7.0 (`HAS_WARP_MERGE_SORT() && USE_ROCM`, for `slice_size <= 256`), so the user-visible regression is ROCm-only. The same kernel is compiled on CUDA builds (`CUDA_VERSION >= 11.6`) but is not currently wired into the CUDA dispatcher; the fix applies there too if it ever is.

## Fix

Gate the sentinel selection on `std::numeric_limits<scalar_t>::has_infinity`:
- floating types keep using `±infinity()` (unchanged behavior),
- integer types use `lowest()` (descending) / `max()` (ascending) so padding always sorts to the tail of the desired ordering and is never written into the top-k output.

## Repro

Single-stream, deterministic, fails 100% on a build that contains #170029 without this fix:

```python
import torch
a = torch.randint(-1 << 30, 1 << 30, (4096, 71), dtype=torch.int64, device='cuda')
v, i = a.topk(71, dim=1)
assert i.min() >= 0 and i.max() < 71  # fails before this patch
```

Externally observed: a customer dump (int32, slice_size=150, k=67, largest=False) reproduces 5000/5000 iterations on a built-from-source ROCm 7.0.2 PyTorch with #170029, and 0/5000 with this patch applied.

## Test plan

- [x] Added `test/test_sort_and_select.py` regression test exercising slice sizes in the `warpMergeSortTopK` dispatch window (≤ 256), all integer dtypes, `k == slice_size`, both `largest=True` and `False`.
- [x] Verified deterministic repro above goes from fail → pass with this patch.
- [x] Verified customer dump (int32, slice=150, k=67) goes from 5000/5000 OOB → 0/5000 OOB on a fresh build.
- [x] Sanity-checked floating dtypes (`float32`, `float64`, `Half`, `BFloat16`) unchanged via the same regression test.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @sarthaktandon9amd

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Review by jeffdaily — APPROVED** [review]

**Comment by haowu14** [comment]
Validated this fix on the workload that originally detected the problem.
